### PR TITLE
Frontend Create Backfill

### PR DESCRIPTION
**Pull request by aLekSer**
Initial version of CreateBackfill. It stores an empty regular TicketIds list.

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:
Wait on StateStare changes.

### CODE REVIEW

File: internal/app/frontend/frontend_service.go [hunk 1]
```diff
@@ -94,6 +94,47 @@ func doCreateTicket(ctx context.Context, req *pb.CreateTicketRequest, store stat
...
+	sfCount += len(backfill.GetSearchFields().GetDoubleArgs())
+	sfCount += len(backfill.GetSearchFields().GetStringArgs())
+	sfCount += len(backfill.GetSearchFields().GetTags())
+	stats.Record(ctx, searchFieldsPerBackfill.M(int64(sfCount)))
+	stats.Record(ctx, totalBytesPerBackfill.M(int64(proto.Size(backfill))))
+
+	err := store.CreateBackfill(ctx, backfill, []string{})
+	if err != nil {
+		return nil, err
+
```
Comment (hsorellana) [inline internal/app/frontend/frontend_service.go:134]: There's a missing closing curly brace here

Comment (aLekSer) [reply internal/app/frontend/frontend_service.go:134]: That's true, I cannot run the test without state store changes.

File: internal/api/internal.proto [hunk 2]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2020 Google LLC
```
Comment (scosgrave) [inline internal/api/internal.proto:1]: I think `statestore.proto` might be a more appropriate name for this file. It matches the style of using the module name in the proto file name that `synchronizer.proto` has started.

Comment (Laremere) [reply internal/api/internal.proto:1]: messages.proto would also be appropriate, as we put all message definitions that aren't request/response types in a similarly named file in the public api.

Comment (Laremere) [reply internal/api/internal.proto:1]: Oh that file is actually being added by https://github.com/googleforgames/open-match/pull/1273

Comment (scosgrave) [reply internal/api/internal.proto:1]: My initial take on how this file should be named was not right. Laramere educated me on how these files are named. The naming pattern is not based on the module name. Instead, services and their req/resp usually go into their own proto file, and objects that aren't coupled to a service's API usually go in a messages.proto file. 
+1 for messages.proto

Comment (aLekSer) [reply internal/api/internal.proto:1]: I will copy a following thread back to the author @akremsa

File: internal/app/frontend/frontend_service.go [hunk 3]
```diff
@@ -94,6 +94,46 @@ func doCreateTicket(ctx context.Context, req *pb.CreateTicketRequest, store stat
...
+func doCreateBackfill(ctx context.Context, req *pb.CreateBackfillRequest, store statestore.Service) (*pb.Backfill, error) {
+	// Generate an id and create a Backfill in state storage
+	backfill, ok := proto.Clone(req.Backfill).(*pb.Backfill)
+	if !ok {
+		return nil, status.Error(codes.Internal, "failed to clone input ticket proto")
+	}
+
+	backfill.Id = xid.New().String()
+	backfill.CreateTime = ptypes.TimestampNow()
```
[the remark this answers is not in the document]
Comment (aLekSer) [reply internal/app/frontend/frontend_service.go:121]: This is an error to have not nil createTime:
```
		return nil, status.Errorf(codes.InvalidArgument, "backfills cannot be created with create time set")
```

File: internal/app/frontend/frontend_service.go [hunk 4]
```diff
@@ -94,6 +94,46 @@ func doCreateTicket(ctx context.Context, req *pb.CreateTicketRequest, store stat
 	return ticket, nil
 }
 
+// CreateBackfill creates a new Backfill object.
+// it assigns an unique Id to the input Backfill and record it in state storage.
+// A Backfill is considered as ready for matchmaking once it is created.
+//   - If SearchFields exist in a Backfill, CreateBackfill will also index these fields such that one can query the ticket with query.QueryBackfills function.
+func (s *frontendService) CreateBackfill(ctx context.Context, req *pb.CreateBackfillRequest) (*pb.Backfill, error) {
```
Comment (akremsa) [inline internal/app/frontend/frontend_service.go:101]: why don't you check req for nil?

File: internal/app/frontend/frontend_service_test.go [hunk 5]
```diff
@@ -274,3 +330,62 @@ func TestDoGetTicket(t *testing.T) {
...
+	for _, test := range tests {
+		test := test
+		t.Run(test.description, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(utilTesting.NewContext(t))
+			store, closer := statestoreTesting.NewStoreServiceForTesting(t, viper.New())
+			defer closer()
+
+			test.preAction(ctx, cancel, store)
+
+			backfill, l, err := store.GetBackfill(ctx, fakeBackfill.GetId())
```
Comment (akremsa) [inline internal/app/frontend/frontend_service_test.go:381]: require.NotNil(t, backfill)

File: internal/testing/e2e/backfill_test.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+		},
+	}}}
+	b1, err := om.Frontend().CreateBackfill(ctx, bf)
+	require.Nil(t, err)
+
+	b2, err := om.Frontend().CreateBackfill(ctx, bf)
+	require.Nil(t, err)
+
+	// Different Ids should be generated
+	assert.NotEqual(t, b1.Id, b2.Id)
```
Comment (akremsa) [inline internal/testing/e2e/backfill_test.go:43]: why do you mix assert and require? 
please use require

File: internal/testing/e2e/backfill_test.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	b1.Id = b2.Id
+	b1.CreateTime = b2.CreateTime
+	// Other than CreateTune abd Id fields, they should be equal
+	assert.Equal(t, b1, b2)
+	require.Nil(t, err)
+	get, err := om.Frontend().GetBackfill(ctx, &pb.GetBackfillRequest{BackfillId: b1.Id})
```
Comment (akremsa) [inline internal/testing/e2e/backfill_test.go:49]: please use common naming like **result** or **actual**, it will be easier to read this code.

File: internal/testing/e2e/backfill_test.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+func TestCreateGetBackfill(t *testing.T) {
+	om := newOM(t)
+	ctx := context.Background()
+
+	bf := &pb.CreateBackfillRequest{Backfill: &pb.Backfill{SearchFields: &pb.SearchFields{
+		StringArgs: map[string]string{
+			"search": "me",
+		},
+	}}}
+	b1, err := om.Frontend().CreateBackfill(ctx, bf)
```
Comment (akremsa) [inline internal/testing/e2e/backfill_test.go:36]: require.NotNil(t, b1)

File: internal/testing/e2e/backfill_test.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+	require.Nil(t, err)
+
+	b2, err := om.Frontend().CreateBackfill(ctx, bf)
```
Comment (akremsa) [inline internal/testing/e2e/backfill_test.go:39]: require.NotNil(t, b2)

File: internal/testing/e2e/backfill_test.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	require.Nil(t, err)
+
+	// Different Ids should be generated
+	assert.NotEqual(t, b1.Id, b2.Id)
+	b1.Id = b2.Id
+	b1.CreateTime = b2.CreateTime
+	// Other than CreateTune abd Id fields, they should be equal
```
Comment (akremsa) [inline internal/testing/e2e/backfill_test.go:46]: What does it mean? Could you rephrase it, please?

File: internal/app/frontend/frontend_service.go [hunk 11]
```diff
@@ -94,6 +94,46 @@ func doCreateTicket(ctx context.Context, req *pb.CreateTicketRequest, store stat
...
+	// Perform input validation.
+	if req.Backfill == nil {
```
Comment (akremsa) [inline internal/app/frontend/frontend_service.go:103]: Please test these req validations in e2e tests

File: internal/app/frontend/frontend_service_test.go [hunk 12]
```diff
@@ -88,6 +88,62 @@ func TestDoCreateTickets(t *testing.T) {
...
+	for _, test := range tests {
+		test := test
+		t.Run(test.description, func(t *testing.T) {
+			store, closer := statestoreTesting.NewStoreServiceForTesting(t, cfg)
+			defer closer()
+
+			ctx, cancel := context.WithCancel(utilTesting.NewContext(t))
+			test.preAction(cancel)
+
+			res, err := doCreateBackfill(ctx, &pb.CreateBackfillRequest{Backfill: test.backfill}, store)
```
Comment (akremsa) [inline internal/app/frontend/frontend_service_test.go:135]: require.NotNil(t, res)

Comment (akremsa) [reply internal/app/frontend/frontend_service_test.go:135]: Why do you test doCreateBackfill instead of CreateBackfill?

**Review by akremsa — CHANGES_REQUESTED** [review]
Some minor changes requested

File: internal/app/frontend/frontend_service.go [hunk 13]
```diff
@@ -94,6 +94,46 @@ func doCreateTicket(ctx context.Context, req *pb.CreateTicketRequest, store stat
...
+	sfCount += len(backfill.GetSearchFields().GetDoubleArgs())
+	sfCount += len(backfill.GetSearchFields().GetStringArgs())
+	sfCount += len(backfill.GetSearchFields().GetTags())
+	stats.Record(ctx, searchFieldsPerBackfill.M(int64(sfCount)))
+	stats.Record(ctx, totalBytesPerBackfill.M(int64(proto.Size(backfill))))
+
+	err := store.CreateBackfill(ctx, backfill, []string{})
+	if err != nil {
+		return nil, err
+	}
```
Comment (hsorellana) [inline internal/app/frontend/frontend_service.go:133]: Before returning, after my PR is merged, here should be a store.IndexBackfill call also

Comment (aLekSer) [reply internal/app/frontend/frontend_service.go:133]: That's expected to have. I will add in a different PR.

File: internal/app/frontend/frontend_service_test.go (hunk 12 above)
Comment (aLekSer) [reply internal/app/frontend/frontend_service_test.go:135]: I am testing real handlers as part of E2E test.

File: internal/testing/e2e/backfill_test.go (hunk 10 above)
Comment (aLekSer) [reply internal/testing/e2e/backfill_test.go:46]: I have updated this comment, but in a separate branch, sorry. I meant that all fields other then CreateTime and Ids are the same if we CreateBackfill with the same input args.

File: internal/app/frontend/frontend_service_test.go [hunk 14]
```diff
@@ -88,6 +88,62 @@ func TestDoCreateTickets(t *testing.T) {
 	}
 }
 
+func TestDoCreateBackfill(t *testing.T) {
```
Comment (Laremere) [inline internal/app/frontend/frontend_service_test.go:91]: While I think this follows the patterns of the testing for the other calls, I'm unsure about the value these tests are adding:

- I don't think calling with canceled context actually finds problems, at most if checks that one use of context is done correctly, but not that all uses are correct.
- It took a bit to get there, but I'm pretty confident that "preAction" is an anti-pattern at this point.  It's adding too much logic to tests.
- See also, the if err != nil check in the test code is definitely logic.  This should be removed, and is evidence of the tests being over-generalized.

My recommendation:
Get the test coverage using the e2e tests for:
- creating
- getting backfill created
- getting backfill which isn't present, and getting not found
(first two are already done, as far as I can tell)

At most, keep only the canceled context tests here, but I'm fine not having those.
The other tests that follow these patterns for existing features should probably be cleaned up, but that's out of scope for this project.

Comment (aLekSer) [reply internal/app/frontend/frontend_service_test.go:91]: Thanks for your feedback, will address all comments

File: internal/app/frontend/frontend_service_test.go (hunk 12 above)
Comment (akremsa) [reply internal/app/frontend/frontend_service_test.go:135]: This is a unit test, it has nothing to do with e2e one. _doCreateBackfill_ is not an entry point to the current package. Please, test exported methods only, as you do in TestDoGetBackfill.

Comment (aLekSer) [reply internal/app/frontend/frontend_service_test.go:135]: First of all we follow the same pattern for `TestDoCreateTicket()`, second Scott asked me to use E2E to cover those handlers:
https://github.com/googleforgames/open-match/pull/1279/files#r528961563

Comment (akremsa) [reply internal/app/frontend/frontend_service_test.go:135]: If there is the same approach in TestDoCreateTicket it doesn't mean that it is done right. You should not rely on any other tests (e2e, etc) while working on unit tests. Here you **ignore** everything which comes before _doCreateBackfill_.

Comment (akremsa) [reply internal/app/frontend/frontend_service_test.go]: @Laremere do you mean that we should test public API in e2e tests, rather than unit ones?
Because at this moment I don't see a clear boundary between tickets e2e and tickets unit tests, for instance.
If so - there is no point to add and support tests in _frontend_ package right now.

Comment (akremsa) [reply internal/app/frontend/frontend_service_test.go]: With _TestDoCreateTickets_ half of a method is tested in frontend package, another part (validations) in e2e, which is a bit confusing.

**Review by akremsa — APPROVED** [review]

File: internal/app/frontend/frontend_service_test.go [hunk 15]
```diff
@@ -88,6 +89,97 @@ func TestDoCreateTickets(t *testing.T) {
...
+						}}}},
+			expectedCode:    codes.OK,
+			expectedMessage: "",
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.description, func(t *testing.T) {
+			res, err := fs.CreateBackfill(ctx, tc.request)
```
Comment (akremsa) [inline internal/app/frontend/frontend_service_test.go:152]: Ok, now it's better. But why do you retain TestDoCreateTickets now?

**Comment by aLekSer** [comment]
Hello @Laremere I have applied all comments, if I did something wrong please leave feedback.

File: internal/app/frontend/frontend_service_test.go [hunk 16]
```diff
@@ -88,6 +89,97 @@ func TestDoCreateTickets(t *testing.T) {
...
+			if tc.expectedCode == codes.OK {
```
Comment (Laremere) [inline internal/app/frontend/frontend_service_test.go:153]: Avoid clauses like this in tests: instead this should be two separate test cases.

File: internal/app/frontend/frontend_service_test.go [hunk 17]
```diff
@@ -88,6 +89,97 @@ func TestDoCreateTickets(t *testing.T) {
...
+				require.NotNil(t, res)
+			} else {
+				require.Error(t, err)
+				require.Equal(t, tc.expectedCode.String(), status.Convert(err).Code().String())
+				require.Contains(t, status.Convert(err).Message(), tc.expectedMessage)
+			}
+		})
+	}
+
+	// expect error with canceled context
```
Comment (Laremere) [inline internal/app/frontend/frontend_service_test.go:164]: Should be in a separate test case.

File: internal/testing/e2e/backfill_test.go [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+	// All fields other than CreateTime and Id fields should be equal
+	require.Equal(t, b1, b2)
+	require.Nil(t, err)
+	actual, err := om.Frontend().GetBackfill(ctx, &pb.GetBackfillRequest{BackfillId: b1.Id})
+	require.Nil(t, err)
+	require.NotNil(t, actual)
+	require.Equal(t, b1, actual)
+
+	// Get Backfill which is not present - NotFound
+	actual, err = om.Frontend().GetBackfill(ctx, &pb.GetBackfillRequest{BackfillId: b1.Id + "new"})
```
Comment (Laremere) [inline internal/testing/e2e/backfill_test.go:64]: I think this should be a separate test case?

**Review by Laremere — APPROVED** [review]
Adding a few comments, don't want to block this on nit for testing, so merging with comments.  Cleanup to match my desired state doesn't involve only the new tests here, so it's fine.

File: internal/testing/e2e/backfill_test.go (hunk 18 above)
Comment (aLekSer) [reply internal/testing/e2e/backfill_test.go:64]: OK, I will make it as a separate PR.
